pydra/tasks/fitlins/utils/bids.py

```diff
@@ -0,0 +1,173 @@
+"""
+Utilities to handle BIDS inputs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+ADD EXAMPLES
+"""
+import os
+import json
+import warnings
+import copy
+from itertools import chain
+
+
+class BIDSError(ValueError):
+    def __init__(self, message, bids_root):
+        indent = 10
+        header = '{sep} BIDS root folder: "{bids_root}" {sep}'.format(
+            bids_root=bids_root, sep=''.join(['-'] * indent)
+        )
+        self.msg = '\n{header}\n{indent}{message}\n{footer}'.format(
+            header=header,
+            indent=''.join([' '] * (indent + 1)),
+            message=message,
+            footer=''.join(['-'] * len(header)),
+        )
+        super(BIDSError, self).__init__(self.msg)
+        self.bids_root = bids_root
+
+
+class BIDSWarning(RuntimeWarning):
+    pass
+
+
+def load_all_specs(all_specs, specs, node, **filters):
+    if node.level == 'run':
+        specs = node.run(group_by=node.group_by, force_dense=False)
+    else:
+        contrasts = list(chain(*[s.contrasts for s in specs]))
+        specs = node.run(contrasts, group_by=node.group_by, **filters)
+
+    all_specs[node.name] = specs
+    for child in node.children:
+        load_all_specs(all_specs, specs, child.destination, **child.filter)
+
+
+def collect_participants(layout, participant_label=None, strict=False):
+    """
+    List the participants under the BIDS root and checks that participants
+    designated with the participant_label argument exist in that folder.
+    Returns the list of participants to be finally processed.
+    Requesting all subjects in a BIDS directory root:
+    >>> collect_participants(layout)
```

Comment:
hello @effigies I keep getting errors here from the test, saying that `NameError: name 'layout' is not defined`. I don't fully understand why those lines as comments will be examined in the test. I guess if I remove those lines, the error should be resolved?

Reply:
oh I got it, it's doctest....

Reply:
I don't think fitlins depends on niworkflows, and this package definitely shouldn't.

Reply:
I think removing broken doctests for now is probably the best plan.

Reply:
Thank you and sorry I just saw this! I'll fix it soon. 
Also, I just get a better understanding of how to use `typing` represent `traits.Enum/Either`. I'll fix some of that too!

Reply:
fixed this one (and we have other new problems) so I'm gonna resolve this conversation